src/erlazure_insights.erl

```diff
@@ -0,0 +1,130 @@
+            {ok, Data};
+        {ok, Data, NextToken} ->
+            follow_token(binary_to_list(NextToken), [DataAcc | Data], Config);
+        {error, _Reason} = ErrorCase ->
+            {ErrorCase, {accumulated, DataAcc}}
+    end.
+
+request_api(Url, Config) ->
+    Response = httpc:request(get, {Url, construct_headers(Config)}, [], []),
+
+    case Response of
+        {ok, {{_Version, 200, _Phrase}, _Headers, Body}} ->
+            DecodedBody = jsx:decode(list_to_binary(Body)),
+            NextLink = proplists:get_value(<<"nextLink">>, DecodedBody, undefined),
+            Values = proplists:get_value(<<"value">>, DecodedBody, undefined),
+            {ok, Values, NextLink};
+        {ok, {{_Version, Code, Phrase}, _Headers, Body}} ->
+            {error, {Code, {Phrase, Body}}};
+        {error, _Reason} = ErrorCase ->
+            ErrorCase
+    end.
+
+http_request(SubscriptionId, Filter, Config) ->
+    URL = "https://management.azure.com/subscriptions/" ++
+    SubscriptionId ++
+    "/providers/microsoft.insights/eventtypes/management/values?" ++ "api-version=2014-04-01" ++ "&" ++ Filter,
+    request_api(URL, Config).
+
+-spec get_time_offset_as_timestamp(datetime(), offset()) -> string().
+get_time_offset_as_timestamp(DateTime = {_Date, _Time}, {Unit, Offset}) ->
+
+    Seconds = calendar:datetime_to_gregorian_seconds(DateTime),
+
+    NewDateTime = case Unit of
+        seconds -> calendar:gregorian_seconds_to_datetime(Seconds - Offset);
+        minutes -> calendar:gregorian_seconds_to_datetime(Seconds - Offset * 60);
+        hours   -> calendar:gregorian_seconds_to_datetime(Seconds - Offset * 60 * 60)
+    end,
+
+    erlazure_utils:iso_8601_fmt(NewDateTime).
+
+create_time_bounded_filter(TS1, TS2) ->
+    Filter = "$filter=" ++ http_uri:encode("eventTimestamp ge '" ++
+        TS1  ++
+    "' and eventTimestamp le '" ++
+    TS2 ++ "'"),
+    Filter.
+
+construct_headers(Config) ->
+    [{"Authorization", "Bearer " ++ Config#azure_config.auth_token},
+        {"Accept", "application/json"}].
+
+-ifdef(TEST).
+
+test_url_construction() ->
+    SubId = "SubId",
+    FromTimestamp = "FromTimestamp",
+    ToTimestamp = "ToTimestamp",
+    Config = #azure_config{auth_token="Token"},
+
+    meck:new(httpc, []),
+    meck:expect(httpc, request, 
+        fun(get, {Url, _}, _, _) -> 
+                TestUrl = "https://management.azure.com/subscriptions/" ++ 
+                SubId ++ 
+                "/providers/microsoft.insights/eventtypes/management/values?api-version=2014-04-01&" ++
+                "$filter=" ++ 
+                http_uri:encode("eventTimestamp ge '" ++ 
+                    FromTimestamp ++ 
+                    "' and eventTimestamp le '" 
+                    ++ ToTimestamp ++ "'"),
+
+                true = (TestUrl == Url),
+                {ok, {{"version", 401, "Phrase"}, "Headers", "Body"}}
+        end
+    ),
+
+    http_request(SubId, FromTimestamp, ToTimestamp, Config),
```

Comment:
I think that high level functions are better tested in a common test environment, since you are testing overall behaviour rather than individual function logic. The ```http_request``` function is a nice one to test along with the ```get_time_offset_as_timestamp``` and ```create_time_bounded_filter```because they do 'one thing' where as get_events* functions do a lot of things that you really need to test together in a common test environment